// @vue/language-core/lib/codegen/names.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StyleModules = exports.PublicProps = exports.SetupExposed = exports.Emit = exports.EmitProps = exports.ModelEmit = exports.RootEl = exports.TemplateRefs = exports.InheritedAttrs = exports.Slots = exports.Props = exports.ModelProps = exports._export = exports.intrinsics = exports.directives = exports.components = exports.setup = exports.exposed = exports.emit = exports.modelEmit = exports.defaultModels = exports.defaults = exports.props = exports.slots = exports.dollars = exports.ctx = void 0;
exports.ctx = '__VLS_ctx';
exports.dollars = '__VLS_dollars';
exports.slots = '__VLS_slots';
exports.props = '__VLS_props';
exports.defaults = '__VLS_defaults';
exports.defaultModels = '__VLS_defaultModels';
exports.modelEmit = '__VLS_modelEmit';
exports.emit = '__VLS_emit';
exports.exposed = '__VLS_exposed';
exports.setup = '__VLS_setup';
exports.components = '__VLS_components';
exports.directives = '__VLS_directives';
exports.intrinsics = '__VLS_intrinsics';
exports._export = '__VLS_export';
exports.ModelProps = '__VLS_ModelProps';
exports.Props = '__VLS_Props';
exports.Slots = '__VLS_Slots';
exports.InheritedAttrs = '__VLS_InheritedAttrs';
exports.TemplateRefs = '__VLS_TemplateRefs';
exports.RootEl = '__VLS_RootEl';
exports.ModelEmit = '__VLS_ModelEmit';
exports.EmitProps = '__VLS_EmitProps';
exports.Emit = '__VLS_Emit';
exports.SetupExposed = '__VLS_SetupExposed';
exports.PublicProps = '__VLS_PublicProps';
exports.StyleModules = '__VLS_StyleModules';
