// @jsonjoy.com/json-pack/lib/nfs/v3/mount/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MountExportResponse = exports.MountExportRequest = exports.MountUmntallRequest = exports.MountUmntRequest = exports.MountDumpResponse = exports.MountDumpRequest = exports.MountMntResponse = exports.MountMntResOk = exports.MountMntRequest = void 0;
class MountMntRequest {
    constructor(dirpath) {
        this.dirpath = dirpath;
    }
}
exports.MountMntRequest = MountMntRequest;
class MountMntResOk {
    constructor(fhandle, authFlavors) {
        this.fhandle = fhandle;
        this.authFlavors = authFlavors;
    }
}
exports.MountMntResOk = MountMntResOk;
class MountMntResponse {
    constructor(status, mountinfo) {
        this.status = status;
        this.mountinfo = mountinfo;
    }
}
exports.MountMntResponse = MountMntResponse;
class MountDumpRequest {
}
exports.MountDumpRequest = MountDumpRequest;
class MountDumpResponse {
    constructor(mountlist) {
        this.mountlist = mountlist;
    }
}
exports.MountDumpResponse = MountDumpResponse;
class MountUmntRequest {
    constructor(dirpath) {
        this.dirpath = dirpath;
    }
}
exports.MountUmntRequest = MountUmntRequest;
class MountUmntallRequest {
}
exports.MountUmntallRequest = MountUmntallRequest;
class MountExportRequest {
}
exports.MountExportRequest = MountExportRequest;
class MountExportResponse {
    constructor(exports) {
        this.exports = exports;
    }
}
exports.MountExportResponse = MountExportResponse;
